From 1ab013251b1399ce6d04cbce83f4638c6124a822 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@apache.org>
Date: Tue, 30 Aug 2016 14:09:14 +0000
Subject: [PATCH 1798/2748] MAPREDUCE-4784. TestRecovery occasionally fails.
 Contributed by Haibo Chen (cherry picked from
 commit af508605a9edc126c170160291dbc2fe58b66dea)

Conflicts:

	hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/TestRecovery.java

Change-Id: I653eb0e738c8c8454415acd96606056361467d2a
---
 .../hadoop/mapreduce/v2/app/TestRecovery.java      |   26 +++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/TestRecovery.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/TestRecovery.java
index 713fd21..441a3b9 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/TestRecovery.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-app/src/test/java/org/apache/hadoop/mapreduce/v2/app/TestRecovery.java
@@ -25,6 +25,8 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+
+import com.google.common.base.Supplier;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
@@ -34,6 +36,7 @@
 import java.util.List;
 import java.util.Map;
 
+import java.util.concurrent.TimeoutException;
 import org.junit.Assert;
 
 import org.apache.commons.logging.Log;
@@ -91,6 +94,7 @@
 import org.apache.hadoop.mapreduce.v2.util.MRBuilderUtils;
 import org.apache.hadoop.security.Credentials;
 import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.test.GenericTestUtils;
 import org.apache.hadoop.util.ReflectionUtils;
 import org.apache.hadoop.yarn.api.records.ApplicationId;
 import org.apache.hadoop.yarn.api.records.ContainerId;
@@ -178,7 +182,10 @@ public void testCrashed() throws Exception {
     Iterator<TaskAttempt> itr = mapTask1.getAttempts().values().iterator();
     itr.next();
     TaskAttempt task1Attempt2 = itr.next();
-    
+
+    // wait for the second task attempt to be assigned.
+    waitForContainerAssignment(task1Attempt2);
+
     // This attempt will automatically fail because of the way ContainerLauncher
     // is setup
     // This attempt 'disappears' from JobHistory and so causes MAPREDUCE-3846
@@ -316,6 +323,21 @@ public void testCrashed() throws Exception {
   }
 
   /**
+   * Wait for a task attempt to be assigned a container to.
+   * @param task1Attempt2 the task attempt to wait for its container assignment
+   * @throws TimeoutException if times out
+   * @throws InterruptedException if interrupted
+   */
+  public static void waitForContainerAssignment(final TaskAttempt task1Attempt2)
+      throws TimeoutException, InterruptedException {
+    GenericTestUtils.waitFor(new Supplier<Boolean>() {
+      @Override public Boolean get() {
+        return task1Attempt2.getAssignedContainerID() != null;
+      }
+    }, 10, 10000);
+  }
+
+  /**
    * AM with 3 maps and 0 reduce. AM crashes after the first two tasks finishes
    * and recovers completely and succeeds in the second generation.
    * 
@@ -1196,6 +1218,8 @@ public void testSpeculative() throws Exception {
     TaskAttempt task1Attempt2 = t1it.next();
     TaskAttempt task2Attempt = mapTask2.getAttempts().values().iterator().next();
 
+    // wait for the second task attempt to be assigned.
+    waitForContainerAssignment(task1Attempt2);
     ContainerId t1a2contId = task1Attempt2.getAssignedContainerID();
 
     LOG.info(t1a2contId.toString());
-- 
1.7.9.5

